Mathlib/Data/Finsupp/MonomialOrder.lean

```diff
@@ -0,0 +1,411 @@
+  add_le_add_left a b h c := by simpa only [add_le_add_iff_left] using h
+
+theorem Finsupp.lex_lt_iff {α N : Type*} [LinearOrder α] [LinearOrder N] [Zero N]
+    {a b : Lex (α →₀ N)} :
+    a < b ↔ ∃ i, (∀ j, j< i → ofLex a j = ofLex b j) ∧ ofLex a i < ofLex b i :=
+    Finsupp.lex_def
+
+theorem Finsupp.lex_le_iff {α N : Type*} [LinearOrder α] [LinearOrder N] [Zero N]
+    {a b : Lex (α →₀ N)} :
+    a ≤ b ↔ a = b ∨ ∃ i, (∀ j, j< i → ofLex a j = ofLex b j) ∧ ofLex a i < ofLex b i := by
+    rw [le_iff_eq_or_lt, Finsupp.lex_lt_iff]
+
+/- -- #check Finsupp.toLex_monotone
+theorem _root_.Finsupp.toLex_monotone' {σ : Type*} [LinearOrder σ] :
+    Monotone (toLex (α := σ →₀ ℕ)) := by
+  apply Finsupp.toLex_monotone
+  intro a b h
+  rw [← (add_tsub_cancel_of_le h), toLex_add]
+  simp only [AddEquiv.refl_symm, le_add_iff_nonneg_right, ge_iff_le]
+  apply bot_le
+-/
+
+variable {σ : Type*} [LinearOrder σ]
+
+/-- The lexicographic order on `σ →₀ ℕ`, as a `MonomialOrder` -/
+noncomputable def MonomialOrder.lex [WellFoundedGT σ] :
+    MonomialOrder σ where
+  syn := Lex (σ →₀ ℕ)
+  locacm := Lex.linearOrderedCancelAddCommMonoid
+  toSyn := {
+    toEquiv := toLex
+    map_add' := toLex_add } -- AddEquiv.refl _ -- (AddEquiv.refl (Lex (σ →₀ ℕ))).symm
+  wf := Lex.wellFoundedLT
+  toSyn_monotone := Finsupp.toLex_monotone
+
+theorem MonomialOrder.lex_le_iff [WellFoundedGT σ] {c d : σ →₀ ℕ} :
+    c ≼[lex] d ↔ toLex c ≤ toLex d := Iff.rfl
+
+theorem MonomialOrder.lex_lt_iff [WellFoundedGT σ] {c d : σ →₀ ℕ} :
+    c ≺[lex] d ↔ toLex c < toLex d := Iff.rfl
+
+/-- The reverse lexicographic order on `σ →₀ ℕ`, as a `MonomialOrder` -/
+noncomputable def MonomialOrder.revLex [WellFoundedLT σ] :
+    MonomialOrder σ :=
+  MonomialOrder.lex (σ := σᵒᵈ)
+
+theorem MonomialOrder.revLex_le_iff [WellFoundedLT σ] {c d : σ →₀ ℕ} :
+    c ≼[revLex] d ↔ (toLex c : Lex (σᵒᵈ →₀ ℕ)) ≤ (toLex d : Lex (σᵒᵈ →₀ ℕ)) :=
+  Iff.rfl
+
+theorem MonomialOrder.revLlex_lt_iff [WellFoundedLT σ] {c d : σ →₀ ℕ} :
+    c ≺[revLex] d ↔ (toLex c : Lex (σᵒᵈ →₀ ℕ)) < (toLex d : Lex (σᵒᵈ →₀ ℕ)) :=
+  Iff.rfl
+
+end Lex
+
+section degLex
+
+/-- A type synonym to equip a type with its lexicographic order sorted by degrees. -/
+def DegLex (α : Type*) := α
+
+variable {α : Type*}
+/-- `toDegLex` is the identity function to the `DegLex` of a type.  -/
+@[match_pattern]
+def toDegLex : α ≃ DegLex α :=
+  Equiv.refl _
+
+/-- `ofDegLex` is the identity function from the `DegLex` of a type.  -/
+@[match_pattern]
+def ofDegLex : DegLex α ≃ α :=
+  Equiv.refl _
+
+@[simp]
+theorem toDegLex_symm_eq : (@toDegLex α).symm = ofDegLex :=
+  rfl
+
+@[simp]
+theorem ofDegLex_symm_eq : (@ofDegLex α).symm = toDegLex :=
+  rfl
+
+@[simp]
+theorem toDegLex_ofDegLex (a : DegLex α) : toDegLex (ofDegLex a) = a :=
+  rfl
+
+@[simp]
+theorem ofDegLex_toDegLex (a : α) : ofDegLex (toDegLex a) = a :=
+  rfl
+
+theorem toDegLex_inj {a b : α} : toDegLex a = toDegLex b ↔ a = b :=
+  Iff.rfl
```

Comment:
I would suggest this:
```lean
theorem toDegLex_injective : Function.Injective (toDegLex (α := α)) := fun _ _ ↦ _root_.id
```

Then, in one direction, you may `apply toDegLex_injective`, and due do the `Function.Injective` API, the `iff` lemma is ` toDegLex_injective.eq_iff`:
```lean
theorem toDegLex_inj {a b : α} : toDegLex a = toDegLex b ↔ a = b := by
  rw [toDegLex_injective.eq_iff]
```
Then, these phrasings `toDegLex_inj` would become unnecessary.

Reply:
I will add `toDegLex_injective`, but I have the impression that `toDegLex_inj` remains useful. At least, this duplication already happens in mathlib, see docs#Subtype.coe_inj and docs#Subtype.coe_injective.